source/fuzz/fuzzer_pass_add_local_variables.cpp

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/fuzzer_pass_add_local_variables.h"
+
+#include "source/fuzz/fuzzer_util.h"
+#include "source/fuzz/transformation_add_local_variable.h"
+#include "source/fuzz/transformation_add_type_pointer.h"
+
+namespace spvtools {
+namespace fuzz {
+
+FuzzerPassAddLocalVariables::FuzzerPassAddLocalVariables(
+    opt::IRContext* ir_context, FactManager* fact_manager,
+    FuzzerContext* fuzzer_context,
+    protobufs::TransformationSequence* transformations)
+    : FuzzerPass(ir_context, fact_manager, fuzzer_context, transformations) {}
+
+FuzzerPassAddLocalVariables::~FuzzerPassAddLocalVariables() = default;
+
+void FuzzerPassAddLocalVariables::Apply() {
+  auto base_type_ids_and_pointers =
+      GetAvailableBaseTypesAndPointers(SpvStorageClassFunction);
+  auto& base_type_ids = base_type_ids_and_pointers.first;
+  auto& base_type_to_pointer = base_type_ids_and_pointers.second;
+
+  for (auto& function : *GetIRContext()->module()) {
+    while (GetFuzzerContext()->ChoosePercentage(
+        GetFuzzerContext()->GetChanceOfAddingLocalVariable())) {
+      uint32_t base_type_id =
+          base_type_ids[GetFuzzerContext()->RandomIndex(base_type_ids)];
+      uint32_t pointer_type_id;
+      std::vector<uint32_t>& available_pointers =
+          base_type_to_pointer.at(base_type_id);
+      if (available_pointers.empty()) {
+        pointer_type_id = GetFuzzerContext()->GetFreshId();
+        available_pointers.push_back(pointer_type_id);
```

Comment:
Might want to add a comment that this modifies the base_type_to_pointer map. This ensure that future loop iterations know that this type exists; it is not useful for this iteration. Could also rename it to available_pointers_to_base_type.

Reply:
Good idea - done (and similar in fuzzer_pass_add_global_variables.cpp)